Codec/Archive/Tar/Check/Internal.hs

```diff
@@ -79,11 +79,12 @@ checkEntrySecurity entry = check (entryPath entry) <|> case entryContent entry o
     _                 -> Nothing
 
   where
+    entryPathUnix = fromTarPathToPosixPath . entryTarPath
     check name
-      | FilePath.Native.isAbsolute name
+      | FilePath.Posix.isAbsolute name
```

Comment:
The description of `checkSecurity` says "These checks are from the perspective of the current OS". The problem is that `c:/foo` is a valid relative path on Linux, but an absolute path on Windows. 

I'm not a fan of the prescribed semantics though: it feels lame to claim that an archive with `c:/foo` poses no security risk just because the check was executed on a Linux machine. Shall we check security in both interpretations, Windows and Posix?

Reply:
My opinion here is this:

- `Entry` as a datatype describes a subset of the spec
- the spec clearly demands unix semantics
- as such all filepaths inside `Entry` must be assumed to be unix filepaths (and indeed, there's no way for `tar` to deal with something like `C:\\Users`)

Hence, we need to enforce unix semantics whenever we create an `Entry`. In this patch, I simply fail at Entry creation time if the native filepath is absolute.

I think this is reasonable.

Reply:
My point was not to change semantics (or relax) `checkEntrySecurity`, it checks paths from the viewpoint of a `Native` system. It's perfectly reasonable to demand that all paths are valid on POSIX, even if packed on Windows, let's just check this in a separate function.

Reply:
Ah, I see what you mean.

But there are a couple of problems regardless. `unpack` does not consistently use the unix->windows conversion functions, but instead assumes:

- `checkEntrySecurity` already failed on invalid paths (but it checked those only post-conversion!)
- the fact that windows supports both path separators is enough 

This seems incorrect/dangerous. 

Further, I believe current `checkEntrySecurity` has a couple of bugs, e.g. it will accept `C:foo`, which is technically a relative path with a drive, but depends on PWD. As such it may point outside the tar root.

So I think we need to:

- convert to native paths consistently, both in `unpack` and `checkEntrySecurity`
- check that all entries are valid unix paths
- be mindful about the platform assumptions

----

E.g. your current PR does this:

https://github.com/haskell/tar/blob/e52c6ec33caa931d273be7a9ab5c41b7367358ef/Codec/Archive/Tar/Check/Internal.hs#L71-L72

Here it starts interpreting the tar path as a native one without prior conversion. That may mostly work, because windows supports both path separators, but I find it questionable. 

The path conversion to native should probably only happen in `check` (or another `checkNative`).